Enum values completion provider for C#

MaStr11 opened the pull request:
Fixes #47193

![xu1IxT2zZ1](https://user-images.githubusercontent.com/24472128/100364808-e5e32780-2ffe-11eb-9768-07c9522d474c.gif)

This PR basically brings the VB version of the enum values provider to C# by moving the existing logic to `AbstractEnumCompletionProvider`.

MaStr11 commented:
@CyrusNajmabadi Is this provider considered useful for C#?

MaStr11 commented:
@sharwell: I talked to @CyrusNajmabadi briefly about the PR and he wanted me to take you into the loop, so you can provide feedback about the user experience requirements. For code review, it is a good idea to view the first commits *Move ... to AbstractEnumCompletionProvider* commit by commit. It shows my porting of the VB specific provider to a language-neutral provider.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/EnumCompletionProvider.cs:
```diff
@@ -0,0 +1,56 @@
+        internal override bool IsInsertionTrigger(SourceText text, int characterPosition, OptionSet options)
+        {
+            var ch = text[characterPosition];
+            return ch is ' ' or '(' or '=' or ':' ||
```
```suggestion
            return this.TriggerCharacters.Contains(ch) ||
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/EnumCompletionProvider.cs:
```diff
@@ -0,0 +1,56 @@
+        protected override string GetInsertionText(CompletionItem item, char ch)
+            => SymbolCompletionItem.GetInsertionText(item);
+
+        protected override (string displayText, string suffix, string insertionText) GetDefaultDisplayAndSuffixAndInsertionText(ISymbol symbol,
+            SyntaxContext context)
+            => CompletionUtilities.GetDisplayAndSuffixAndInsertionText(symbol, context);
```
```suggestion
        protected override (string displayText, string suffix, string insertionText) GetDefaultDisplayAndSuffixAndInsertionText(ISymbol symbol, SyntaxContext context)
            => CompletionUtilities.GetDisplayAndSuffixAndInsertionText(symbol, context);
```

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
```
```suggestion
            if (enumType?.TypeKind != TypeKind.Enum)
```

CyrusNajmabadi replied:
i'm not opposed to patterns.  but when there is already a clear, idiomatic, shorter non-pattern form, i would prefer that.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            // We'll want to build a list of the actual enum members and all accessible instances of that enum, too
+            var result = enumType.GetMembers().Where(m =>
+                m.Kind == SymbolKind.Field &&
+                ((IFieldSymbol)m).IsConst &&
+                m.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation)).ToImmutableArray();
```
```suggestion
            var result = enumType.GetMembers().OfType<IFieldSymbol>().Where(f =>
                f.IsConst && f.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation)).ToImmutableArray();
```

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            return Task.FromResult(result);
+        }
+
+        protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(SyntaxContext context, int position,
+            OptionSet options, CancellationToken cancellationToken)
```
```suggestion
        protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(
            SyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
```

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var syntaxFacts = context.GetLanguageService<ISyntaxFactsService>();
+            var node = syntaxFacts.FindTokenOnLeftOfPosition(context.SyntaxTree.GetRoot(cancellationToken), position, includeSkipped: true).Parent;
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken) ||
+                syntaxFacts.IsSkippedTokensTrivia(node)) // Note: IsSkippedTokensTrivia implementation is probably buggy
```
what is this note about?

CyrusNajmabadi replied:
do we need this check?  do some tests fail if you get rid of it?

MaStr11 replied:
It is a relict from the port. The `IsSkippedTokensTrivia` can be removed without any influence on the tests.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
```
doc.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var span = new TextSpan(position, 0);
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
```
```suggestion
            if (enumType?.TypeKind != TypeKind.Enum)
```

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            }
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var span = new TextSpan(position, 0);
```
what is this for?  it isn't used here.  can you move to where it is used?

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+                s.MatchesKind(SymbolKind.Field, SymbolKind.Local, SymbolKind.Parameter, SymbolKind.Property) &&
+                s.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation));
+
+            var otherInstances = otherSymbols.WhereAsArray(s => Equals(enumType, s.GetSymbolType()));
```
i don't ynderstand what this code is doing.  can you doc why we need this?

MaStr11 replied:
Me neither. This is a 100% code port from the existing VB provider. If I remove `otherInstances` these tests fail:
```VB
        <Fact, Trait(Traits.Feature, Traits.Features.Completion)>
        <WorkItem(815963, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/815963")>
        Public Async Function TestLocalNoAs() As Task
            Dim markup = <Text><![CDATA[
Enum E
    A
End Enum
 
Class C
    Sub M()
        Const e As E = e$$
    End Sub
End Class
]]></Text>.Value
            Await VerifyItemExistsAsync(markup, "e")
            Await VerifyItemIsAbsentAsync(markup, "e As E")
        End Function
```
```CS
        [Fact]
        [Trait(Traits.Feature, Traits.Features.Completion)]
        public async Task TestLocalNoAs()
        {
            var markup = @"
enum E
{
    A
}

class C
{
    public void M()
    {
        const E e = e$$;
    }
}
";

            await VerifyItemExistsAsync(markup, "e");
            await VerifyItemIsAbsentAsync(markup, "e as E");
        }
```
```
    No expected item was found.
    Existing items: E, E.A
    Expected: True
    Actual:   False
```

I have no idea why `e` is considered useful here. I think `otherInstances ` can be removed.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+                        .WithMemberOptions(SymbolDisplayMemberOptions.IncludeContainingType)
+                        .WithLocalOptions(SymbolDisplayLocalOptions.None);
+                    var containingTypeText = symbol.ContainingType.ToMinimalDisplayString(context.SemanticModel, context.Position, displayFormat);
+                    _cachedDisplayAndInsertionText = (symbol.ContainingType, context, containingTypeText);
```
this doesn't seem threadsafe at all.

MaStr11 replied:
Yes. I could wrap it with `ReaderWriterLockSlim` or remove the optimization altogether. What should I do?

CyrusNajmabadi replied:
let's wrap for now.

MaStr11 replied:
I did but used a `lock` because I needed to use `ReaderWriterLockSlim.EnterUpgradeableReadLock` which basically behaves like `lock`.

MaStr11 commented:
@CyrusNajmabadi Thanks for the review. One thing I stumbled upon, is the odd class hierachie especially regarding `CompletionListTag` (green = new classes introduced by this PR):

![Diag](https://user-images.githubusercontent.com/24472128/100451290-6f563080-30b7-11eb-9d87-39654fae6bd9.png)

As you can see there is a mismatch between VB and C#. I haven't looked into `EnumAndCompletionListTagCompletionProvider.cs` so  don't know what it does. I just wanted to bring this to your attention as this feels like a *code smell* to me.

MaStr11 commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+
+            var otherInstances = otherSymbols.WhereAsArray(s => Equals(enumType, s.GetSymbolType()));
+
+            return Task.FromResult(otherInstances.Concat(enumType));
```
The behavior of this method is odd. It only returns the Enum, but not the enum values. For C# I would expect the exact opposite behavior (There are EnumAndCompletionListTagCompletionProvider.cs (and maybe AbstractRecommendationServiceBasedCompletionProvider.cs) that add the Enum type to the completion list).

I can't tell for VB if there is another provider that adds the Enum type.

MaStr11 commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var inferedTypes = typeInferenceService.InferTypes(context.SemanticModel, position, cancellationToken);
```
I changed this call from `InferType` to `InferTypes` as there might be more than one Enum suitable at the position (e.g. method overloads:

https://github.com/dotnet/roslyn/blob/97559e3bd9b136cb2e5f4797b860580f03ae99fd/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumCompletionProviderTests.cs#L567-L598

CyrusNajmabadi replied:
that seemes reasonable.

MaStr11 reviewed (commented):
I added support for pattern combinators in 5125df1, but I'm not really convinced that this is the right way forward. There are probably many more edge cases, where enum values should be suggested, but only the enum type is offered.
Maybe it would be better to integrate this part (suggests the enum values) into the Recommender, which also suggests the enum type.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/EnumCompletionProvider.cs:
```diff
@@ -0,0 +1,55 @@
+                (SyntaxFacts.IsIdentifierStartCharacter(ch) && options.GetOption(CompletionOptions.TriggerOnTypingLetters2, LanguageNames.CSharp));
+        }
+
+        internal override ImmutableHashSet<char> TriggerCharacters { get; } = ImmutableHashSet.Create(' ', '(', '=', ':');
```
are there tests for all of these four cases?

MaStr11 replied:
If `await VerifyItemExistsAsync(markup, "...", usePreviousCharAsTrigger: true);` does such tests, all cases should be covered:

"="
TestRightSideOfAssignment

" "
TestCaseStatement

":"
TestTriggeredForNamedArgument

"("
TestTriggeredOnOpenParen

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+{
+    internal abstract class AbstractEnumCompletionProvider : AbstractSymbolCompletionProvider
+    {
+        private static readonly CompletionItemRules s_rules = CompletionItemRules.Default.WithMatchPriority(MatchPriority.Preselect);
```
how does this work given that the enum type itself will be preselected?

MaStr11 replied:
I don't know (copied from the original VB provider). Should I use `CompletionItemRules.Default` instead? That seems to be the right choice given that the enum type is preselected.

CyrusNajmabadi replied:
i think the two languages have different behaviors.  sounds like C# should have different rules here.

MaStr11 replied:
I played around with this setting in the IDE and at the moment the behavior between C# and VB is the same: The first enum value is pre-selected (even if the enum type is listed above). I think, that is a sensible behavior. I don't have written any tests for that, thought.
These tests should cover the interaction between the SymbolCompletionProvider (offering the Enum type), this provider, pre selection, and item ordering. I will add these tests later.

CyrusNajmabadi replied:
> I played around with this setting in the IDE and at the moment the behavior between C# and VB is the same: The first enum value is pre-selected (even if the enum type is listed above). I think, that is a sensible behavior. I don't have written any tests for that, thought.

i don't think tihs is reasonable as this breaks c# muscle memory.  i.e. in c# i type: 

`Color color = <tab><dot>` and i expect to be given a list of members of `Color`.  With your apporach, the `tab` will complete to the first member which is not how we've worked for the last 15 years.  Please change htis :)

MaStr11 replied:
> Color color = <tab><dot> and i expect to be given a list of members of Color. With your apporach, the tab will complete to the first member which is not how we've worked for the last 15 years. Please change this :)

I totally agree. 

> I played around with this setting in the IDE and at the moment the behavior between C# and VB is the same: The first enum value is pre-selected (even if the enum type is listed above). I think, that is a sensible behavior. I don't have written any tests for that, thought.

I have to take back that statement. I have written some [C# tests now](https://github.com/dotnet/roslyn/pull/49632/files#diff-31303c421816ee9001acbb872ec31e6ef3d701cd04eba6f2a41d02e49d70ae21R2324), that show, that the enum type is still preselected (as it should).

I think in my previous investigation I committed the selection before and so next time the committed enum value was preselected (that is at least what I think happened).

I could still make the `s_rule` provider dependent to be absolutely sure. What do you think?

CyrusNajmabadi replied:
> I could still make the s_rules provider dependent to be absolutely sure. What do you think?

Yup. I would do that

> I think in my previous investigation I committed the selection before and so next time the committed enum value was preselected (that is at least what I think happened).

Can we have a test that verifies that that doesn't happen?  If we commit an enum value, and then invoke completion again, we should still preselect the enum, not the enum value. Thanks!

MaStr11 replied:
> Can we have a test 

I added one in c758cb2

> I could still make the s_rules provider dependent to be absolutely sure.

Did that in 9d0f198

If you look at the test, the enum value is preselected in the second invocation of the completion list. This is true no matter if `GetCompletionItemRules` is overridden or not (Default vs. `WithMatchPriority(MatchPriority.Preselect)`. The test before and after 9d0f198 passes.

I think this is the standard behavior of the completion list: Pre-select the last committed item.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,142 @@
+
+            // This providers provides fully qualified names, eg "DayOfWeek.Monday"
+            // Don't run after dot because SymbolCompletionProvider will provide
+            // members in situations like Dim x = DayOfWeek.$$
```
nice comment, thanks.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -213,6 +214,7 @@ protected override IEnumerable<TypeInferenceInfo> InferTypesWorker_DoNotCallDire
                     AttributeTargetSpecifierSyntax attributeTargetSpecifier => InferTypeInAttributeTargetSpecifier(attributeTargetSpecifier, token),
                     AwaitExpressionSyntax awaitExpression => InferTypeInAwaitExpression(awaitExpression, token),
                     BinaryExpressionSyntax binaryExpression => InferTypeInBinaryOrAssignmentExpression(binaryExpression, binaryExpression.OperatorToken, binaryExpression.Left, binaryExpression.Right, previousToken: token),
+                    BinaryPatternSyntax binaryPattern => GetPatternTypesFromOperation(binaryPattern),
```
only fix if necessary for some actual case we care about.  it's ok for these to be out of sync as they're intended for different purposes.

CyrusNajmabadi commented:
@MaStr11 i'm good with this PR as-is.  do you want to make any other changes before i merge?

MaStr11 commented:
> do you want to make any other changes before I merge?

I think there are some problems with this PR before it should be merged:

As mentioned earlier, this is a port of the VB provider into a language-neutral one. As such, I tried to keep it as close to the original as possible, but there are some issues with the VB provider:

### Important issues, that should be addressed

#### Thread safety
https://github.com/dotnet/roslyn/pull/49632#discussion_r531554006

#### Weird behavior around the symbols for completion:

* `GetPreselectedSymbolsAsync` returns the enum fields
* `GetSymbolsAsync` returns the Enum type itself (but not the values).
* Your https://github.com/dotnet/roslyn/pull/49632/files#r542734395 about `CompletionItemRules.Default.WithMatchPriority(MatchPriority.Preselect)` (I  have to admit that I don't know what that rule is supposed to achieve)

This doesn't make much sense to me, even though it seems to work. If you wouldn't mind (or have an idea of how this is working), I would like to play around to see if it can get simplified.

### Some minor issues

#### Duplication of a method in `TypeInferrer`

see https://github.com/dotnet/roslyn/pull/49632#discussion_r533526234

Is it okay 
* to use `IOperation` and 
* provide a new method, even though there already is an existing one, that does about the same?

#### Re-factoring

`GetPreselectedSymbolsAsync` and `GetSymbolsAsync` share a lot of logic and might gone be reduced to one.

#### Removal of suspicious code

https://github.com/dotnet/roslyn/pull/49632/files#r531552437

I don't see what this code is for. It should be removed.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -1493,6 +1516,19 @@ private IEnumerable<TypeInferenceInfo> InferTypeInRecursivePattern(RecursivePatt
                 return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
             }
 
+            private IEnumerable<TypeInferenceInfo> GetPatternTypesFromOperation(PatternSyntax pattern)
```
can you just remove hte old one and just use this if IOp produces better results all around?

MaStr11 replied:
> can you just remove hte old one and just use this if IOp produces better results all around?

I could not completely replace the old GetPatternTypes with the IOperation based one. The old one had some special logic around tuple deconstruction and NRTs. I integrated my `GetPatternTypesFromOperation` instead (as local function). So there is only one `GetPatternTypes` now.

MaStr11 commented:
@CyrusNajmabadi I implemented enum sorting by underlying value. The idea is, that the sequence of enum values often matters (e.g. `None` being the first entry and `All` being the last entry or severity `Info`, `Warning`, `Error`). This would look like this:

![Qv0rvnCvdg](https://user-images.githubusercontent.com/24472128/102232059-f7796a00-3eee-11eb-9177-fcba35aa713f.png)

I added the underlying value as `inlineDescription`. Hwat do you think? Should we keep this?

MaStr11 commented:
`GetPreselectedSymbolsAsync` and `GetSymbolsAsync` were about the same before. I simplified this by delegating from `GetPreselectedSymbolsAsync`  to `GetSymbolsAsync`.

I needed to change a few test cases after that change: 129c079. The tests didn't make any sense to me in the first place, so I think the change in behavior is okay.

MaStr11 commented:
> @MaStr11 i'm good with this PR as-is. do you want to make any other changes before i merge?

@CyrusNajmabadi I'm ready with the PR now. I did some refactorings and added some tests to make sure that the Enum type and the Enum values are always listed close to each other (in VisualBasic/CSharpCompletionCommandHandlerTests). Namespaces and type aliases required some minor changes, but it's all good now.

I removed the code that did the sorting by underlying enum value (https://github.com/dotnet/roslyn/pull/49632#issuecomment-745352921).

A final review would be appreciated, thanks.

MaStr11 commented on src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb:
```diff
@@ -2097,6 +2097,66 @@ End Class</Document>)
+End Namespace
+</Document>)
+                state.SendInvokeCompletionList()
+                Await state.AssertCompletionItemsDoNotContainAny("A.Colors", "Colors")
```
This is different from C#. In C# the Enum type is listed with namespace "A.Colors" (see test above). I think this is also different from what the behavior is today, without this PR. I think today in VB this is listed as "Colors". But I'm not sure what the source of this "Colors" is today, because in VB lots of types from other namespaces are listed, so it might be "IntelliCode" that lists "Colors" today (and will so in the future).

CyrusNajmabadi replied:
IntelliCode does not run in our tests.

Note: we cannot change behavior here.  It is important for muscle memorythat we still preselect the same item that we used to.  thanks!

MaStr11 replied:
I look more closely and I think the behavior today is broken and the behavior of this PR is "okayish" (example uses the unimported `Microsoft.VisualBasic.FileIO` namespace).

### Today

![PNtnkVbJHm](https://user-images.githubusercontent.com/24472128/102513783-f3368380-408b-11eb-8938-8bca6ba56630.gif)

`FieldType` is from an unimported namespace and listed as `FieldType`. `<tab>`ing it, inserts *FieldType*, which is wrong and should be `FileIO.FieldType`

### Now (this PR)

![TQGVTdm6Vx](https://user-images.githubusercontent.com/24472128/102514337-a2735a80-408c-11eb-991d-cb2aadde9417.gif)

`FieldType` is no longer offered, as the test above also shows.

So this PR removes a broken suggestion. Note: This is only true for unimported namespaces. If the enum type is in scope, it is suggested as expected.

CyrusNajmabadi commented:
> @CyrusNajmabadi I implemented enum sorting by underlying value.

Nice.  i like that.

> I added the underlying value as inlineDescription. What do you think? Should we keep this?

I'm skeptical of hte value there (no pun intended).  Why do i care that hte value of this enum member is '3'?  I owuld remove that.

MaStr11 commented:
This integration test causes permanent errors in this PR (Line :
https://github.com/dotnet/roslyn/blob/a4cae91ca9dc0e33b628b91b0a9a88334e6edc6c/src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpRenameFileToMatchTypeRefactoring.cs#L59-L90

In line 81:
```
Assert.Equal() Failure
↓ (pos 0)
Expected: class MismatchedClassName { }
Actual:   public class MismatchedClassName { }
↑ (pos 0)
```

I have no idea how this relates to the changes in this PR and can't debug the test. Any ideas (the only inference I see is the change in the CompletionProvider extension order, but that seems a very unlikely cause)?

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -237,10 +239,12 @@ protected override IEnumerable<TypeInferenceInfo> InferTypesWorker_DoNotCallDire
                     SwitchStatementSyntax switchStatement => InferTypeInSwitchStatement(switchStatement, token),
                     ThrowStatementSyntax throwStatement => InferTypeInThrowStatement(throwStatement, token),
                     TupleExpressionSyntax tupleExpression => InferTypeInTupleExpression(tupleExpression, token),
+                    UnaryPatternSyntax unaryPattern => GetPatternTypes(unaryPattern),
```
are all tehse new cases tested?

CyrusNajmabadi replied:
note: we have explicit tests around the TypeInferenceService, can you updat there?

MaStr11 replied:
I copied the tests from EnumCompletionProviderTests to TypeInferrerTests in 3479b73e0bc76de30601aa547efa66e7f2ae2584, 48ec7934328121de6d761535c4822c7d0b7b9df6 and de23198eeebeb37ca3ed072d48acab99948c6c5a

TestEnumInPatterns_Is_ConstUnaryAndBinaryPattern

https://github.com/dotnet/roslyn/pull/49632/files#diff-9cda451529a635b3dfcee6e983569f6139cdb5be2c64d916dca49bdb37152a2aR3109

> are all tehse new cases tested?

Yes. All code paths are covered by the new tests in TypeInferrerTests.

Tests are duplicated in EnumCompletionProviderTests and TypeInferrerTests with some minor differences now. I think this is valuable, but I can remove the tests from EnumCompletionProviderTests if you like.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -1494,14 +1517,28 @@ private IEnumerable<TypeInferenceInfo> InferTypeInRecursivePattern(RecursivePatt
+                        return CreateResult(resultType);
+                    }
+
+                    return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
```
can you doc what's going on here?

MaStr11 replied:
Done in 934ba760

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -1494,14 +1517,28 @@ private IEnumerable<TypeInferenceInfo> InferTypeInRecursivePattern(RecursivePatt
+                    _ => GetPatternTypesFromOperation(pattern),
                 };
 
+                IEnumerable<TypeInferenceInfo> GetPatternTypesFromOperation(PatternSyntax pattern)
```
i don't see any value in a local function here.  Why not just a normal switch, and then this local function will be what you do in the default case?

MaStr11 replied:
Inlined in 4b67cba2

MaStr11 commented:
@CyrusNajmabadi I did the changes as requested, but this problem https://github.com/dotnet/roslyn/pull/49632#issuecomment-748269230 still applies and I have no idea what's wrong with it and how to address it.

CyrusNajmabadi commented:
@MaStr11 i'll help pushing this in.  I think we should take it for 16.10 instead of 16.9 though as this impacts a core completion scenerio.  That will help us ensure there's enogh time to feel this out and make sure it's not impacting important muscle memory scenarios.  @jinujoseph for visibility.

CyrusNajmabadi commented:
>@CyrusNajmabadi I did the changes as requested, but this problem #49632 (comment) still applies and I have no idea what's wrong with it and how to address it.

Looking into this.  It repros for me.

CyrusNajmabadi commented:
There appears to be an extra entry in the undo stack.  Very weird.  Investigating further.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs:
```diff
@@ -0,0 +1,131 @@
+                }
+            }
+
+            return displayText;
```
why are you doing this?  why was this information in teh display text in the first place?  if it was needed for the display text, why is it not appropriate to spit into the final user code?

CyrusNajmabadi commented:
Ok.  Fixed the issue.  There was an extra entry in the undo stack due to this completion provider showing up in an innapropriate context.  I fixed things up to use the normal C# enum provider that we already have so that we only show these members in contexts where we had already decided to show and select the normal enum type. 

Once this passes, i can merge into 16.10.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:
```diff
@@ -178,24 +180,34 @@ private static Glyph GetGlyph(SymbolKind kind, Accessibility? declaredAccessibil
+                var originalDefinition = namedType.OriginalDefinition;
                 var taskOfTType = compilation.TaskOfTType();
                 var valueTaskType = compilation.ValueTaskOfTType();
                 var lazyOfTType = compilation.LazyOfTType();
```
@MaStr11 can you lmk what the reason is for the changes in this type?

CyrusNajmabadi replied:
ignore.  sorry.  merge issue.

MaStr11 commented on src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:
```diff
@@ -61,102 +66,118 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
+                var index = 0;
+
+                var fields = type.GetMembers().OfType<IFieldSymbol>().Where(f => f.IsConst).Where(f => f.HasConstantValue);
+                foreach (var field in fields.OrderBy(f => IntegerUtilities.ToInt64(f.ConstantValue)))
```
Sorting fails for enums backed by a  `ulong` with values > `Int64.MaxValue`. @CyrusNajmabadi Do you want me to fix this?

CyrusNajmabadi replied:
Feels unlikely to matter in practice. But if you'd like, go ahead.
